docs/source/deep_dives/sample_packing.rst

```diff
@@ -0,0 +1,247 @@
+.. _sample_packing_deepdive:
+
+==========================================================================================
+Fine-tuning in torchtune with sample packing using in-sample position encoding and masking
+==========================================================================================
+
+torchtune now supports sample packing for significant boosts in training throughput. This
+can be enabled on CLI, in the yaml config, or in any dataset builder by setting ``packed=True``.
+
+.. code-block:: python
+
+    from torchtune.datasets import alpaca_dataset, PackedDataset
+
+    # Load in tokenizer
+    tokenizer = ...
+    dataset = alpaca_dataset(
+        tokenizer=tokenizer,
+        packed=True,
+    )
+    print(isinstance(dataset, PackedDataset))  # True
+
+.. code-block:: yaml
+
+    # YAML config
+    dataset:
+      _component_: torchtune.datasets.alpaca_dataset
+      packed: True
+
+.. code-block:: bash
+
+    # Command line
+    tune run full_finetune_single_device --config llama3/8B_full_single_device \
+    dataset.packed=True
+
+This will automatically adjust attention masks to avoid cross-sample attending and
+update position IDs to be relative, within-sample. In this write-up, we explore what
+the expected performance gains you should see are and how in-sample masking and position
+encoding can affect fine-tuned model quality.
+
+.. grid:: 1
+
+    .. grid-item-card:: :octicon:`mortar-board;1em;` Highlights
+
+      * Sample packing concatenates multiple data samples into a single packed
+      sample up to the maximum sequence length to boost model training throughput.
+      * Enabling sample packing for the Stanford Alpaca dataset yielded up to a
+      5x boost in throughput (~350 tokens/sec/gpu -> ~1800 tokens/sec/gpu)
+      * Adding in-sample position encoding and masking to eliminate cross-attending
+      of unrelated samples improved fine-tuned models on hellaswag (+2%) compared
+      to packing without in-sample masking, but showed no statistically significant
+      difference on truthfulqa_mc2.
+
+
+Background
+----------
+Context windows for modern LLMs are ever-increasing, from Llama3 at 8k to
+Gemini 1.5 Pro which is pushing 2 million. In practice, many OSS text datasets
+(such as Stanford Alpaca) consist of mostly short sequence length samples,
+averaging at ~100-1000, which is a small fraction of the full context. This can
+lead to significant padding compute and memory overhead, even if ignored or masked
+out in the loss function. Sample packing minimizes this overhead by jamming as
+many samples as possible in the extra space in the context window. This can massively
+increase throughput for model training, since the model will see more data per step.
+However, there are several caveats to consider when enabling sample packing for a
+transformer decoder model.
```

Comment:
hope you don't mind me commenting here : )

small reformat:

```
However, there are several caveats to consider when enabling sample packing for a
transformer decoder model:

1. When concatenating multiple samples in a single context window, we need a way
to tell the model to not attend to all other previous samples and only attend to
the sample the current token belongs to. This can be achieved with a block causal
mask to prevent cross-contamination.

2. Similarly, for position encodings we need to encode relative position of a
token within its sample instead of absolute position across all samples in a
single context window.

3. Ideally, the block causal mask and relative position encoding is compatible
with flash attention to maximize speed-ups.

```